libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -321,6 +318,81 @@ string YulUtilFunctions::shiftRightFunctionDynamic()
 	});
 }
 
+string YulUtilFunctions::shiftRightSignedFunctionDynamic()
+{
+	string const functionName = "shift_right_signed_dynamic";
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(bits, value) -> result {
+				<?hasShifts>
+					result := sar(bits, value)
+				<!hasShifts>
+					let divisor := exp(2, bits)
+					switch slt(value, 0)
+					case 0 {
+						result := div(value, divisor)
+					}
+					default {
+						result := not(div(not(value), divisor))
+					}
```

Comment:
In old codegen, we did this jump-less...
```
 					let xor_mask := sub(0, slt(value, 0))
					value_to_shift := xor(div(xor(value, xor_mask), divisor), xor_mask)
```

Reply:
Yes, because the old optimizer was not able to cope with this.

Reply:
I don't think the new one will be able to cope with this either.

Reply:
Which situations did we want to optimize apart from constant values?

Reply:
If the new optimizer can determine that the shift amount is positive, this optimizes more nicely, I think.

Reply:
Ah ok, maybe it does not matter...

Reply:
It's always cheaper not to jump, isn't it?
And if ``slt(value, 0)`` is constant, then the inlined-version will optimize just as nicely...

So the non-constant path is faster due to not jumping and the constant path should be the same.

And this optimization is very hard for the optimizer in general - I don't think it'll be able to decide when it makes sense to remove jumps like this...